libdnf/advisory/advisory_query.cpp

```diff
@@ -0,0 +1,437 @@
+    }
+
+    return *this;
+}
+
+AdvisoryQuery & AdvisoryQuery::ifilter_package(const libdnf::rpm::Package & package, sack::QueryCmp cmp_type) {
+    libdnf::rpm::PackageSet ps(&(advisory_sack->base->get_rpm_solv_sack()));
+    ps.add(package);
+    ifilter_packages(ps, cmp_type);
+    return *this;
+}
+
+
+AdvisoryQuery & AdvisoryQuery::ifilter_packages(const libdnf::rpm::PackageSet & package_set, sack::QueryCmp cmp_type) {
+    auto & solv_sack = advisory_sack->base->get_rpm_solv_sack();
+    Pool * pool = solv_sack.p_impl->get_pool();
+    libdnf::rpm::solv::SolvMap filter_result(solv_sack.p_impl->get_nsolvables());
+    std::vector<AdvisoryPackage> pkgs = get_sorted_advisory_packages();
+
+    bool cmp_not = (cmp_type & libdnf::sack::QueryCmp::NOT) == libdnf::sack::QueryCmp::NOT;
+    if (cmp_not) {
+        // Removal of NOT CmpType makes following comparissons easier and more effective
+        cmp_type = cmp_type - libdnf::sack::QueryCmp::NOT;
+    }
+
+    switch (cmp_type) {
+        case libdnf::sack::QueryCmp::EQ:
+            //TODO(amatej): faster EQ specific version (we can compare whole NEVRA)
+        case libdnf::sack::QueryCmp::GT:
+        case libdnf::sack::QueryCmp::LT:
+        case libdnf::sack::QueryCmp::GTE:
+        case libdnf::sack::QueryCmp::LTE: {
+            for (libdnf::rpm::PackageSet::iterator package = package_set.begin(); package != package_set.end();
+                 package++) {
+                Solvable * solvable = libdnf::rpm::solv::get_solvable(pool, (*package).get_id().id);
+                auto low = std::lower_bound(pkgs.begin(), pkgs.end(), *package, AdvisoryPackage::comparePackageNA);
+                while (low != pkgs.end() && low->p_impl.get()->get_name_id() == solvable->name &&
+                       low->p_impl.get()->get_arch_id() == solvable->arch) {
+                    int libsolv_cmp = pool_evrcmp(pool, low->p_impl.get()->get_evr_id(), solvable->evr, EVRCMP_COMPARE);
+                    if (((libsolv_cmp > 0) && ((cmp_type & sack::QueryCmp::GT) == sack::QueryCmp::GT)) ||
+                        ((libsolv_cmp < 0) && ((cmp_type & sack::QueryCmp::LT) == sack::QueryCmp::LT)) ||
+                        ((libsolv_cmp == 0) && ((cmp_type & sack::QueryCmp::EQ) == sack::QueryCmp::EQ))) {
+                        filter_result.add_unsafe((*low).get_advisory_id().id);
+                    }
+                    ++low;
+                }
+            }
+
+        } break;
+        default:
+            throw NotSupportedCmpType("Unsupported CmpType");
+    }
+
+    // Apply filter results to query
+    if (cmp_not) {
+        p_impl->data_map -= filter_result;
+    } else {
+        p_impl->data_map &= filter_result;
+    }
+
+    return *this;
+}
+
+std::vector<AdvisoryPackage> AdvisoryQuery::get_advisory_packages(
+    const libdnf::rpm::PackageSet & package_set, sack::QueryCmp cmp_type) {
+    std::vector<AdvisoryPackage> pkgs = get_sorted_advisory_packages();
```

Comment:
In the filter we use two types of packages (rpm, advisory). I would recommend to add type of package into the name of the variable.